### PR TITLE
fix(tasks): carry the caller timezone into turns that start at task creation

### PR DESCRIPTION
**Pull request by bluefish-08**
Refs #1675 — closes the last transport gap. Addresses finding **G1** from @rogercloud's review of #1712.

**Stacked on #1713 (which is stacked on #1712).** Cut from that branch, so the diff shown here includes their commits until they merge. Own increment: source +32/-1, tests +229/-2.

## Invariant

This PR guarantees only:

When a task-creation request carries a non-blank `timezone`, the turn that task creation itself starts receives it as `context={"timezone": ...}`, which #1709 renders as local time. When the field is absent or blank, the request behaves exactly as it does today and the turn renders UTC.

## Entry points in scope

- `POST /api/widget/chat/task/create` and `POST /api/share/chat/task/create` (workforce opening turns)
- `POST /v1/chat/tasks` (SDK/REST)
- `create_workforce_run` / `create_workforce_run_by_id`
- `public-agent-chat-page.tsx` opening-message submit

## Why #1712 was not enough

#1712 attaches the zone to `type: "chat"` websocket frames. Two paths never send one:

- **Workforce openers.** `public-agent-chat-page.tsx:336` sends the first message inside the create request and deliberately skips `sendMessage`, because `create_workforce_run` already starts that turn — resending it over the websocket would run the turn twice. So the opening question, the exact shape of the reported incident ("how many shifts do we have on tomorrow?"), was answered against UTC, and later turns cannot repair that first answer.
- **SDK/REST callers.** They have no websocket at all.

## How

The downstream pipe already existed and was simply never connected: `schedule_claimed_create_turn` has taken an optional `context` since before this work (`task_orchestrator.py:602`), and both workforce (`workforce_runs.py:730`) and SDK (`v1/tasks.py:636`) call sites just omitted it. So this is a transport wiring change, not new machinery:

```
TaskCreateRequest.timezone / CreateTaskRequest.timezone
      ↓
create_workforce_run(timezone=...) → _NormalizedWorkforceRunRequest.timezone
      ↓
schedule_claimed_create_turn(context={"timezone": ...})   ← param already existed
      ↓
(unchanged) → execute_task(context=) → metadata["request_context"] → metadata["timezone"]
      ↓
(unchanged, #1709) → both prompt clocks render local time
```

`TaskTurnPayload`, the orchestrator internals, the runner, and the injection points are untouched. Both transports converge on the same `metadata["timezone"]`, so a task that opens over HTTP and continues over websocket keeps one consistent zone.

## Known trade-offs

1. **No IANA validation at the schema layer.** #1709 already resolves through `ZoneInfo` and degrades to UTC on anything unresolvable, and the websocket path does not validate either. Validating here would make the same untrusted field 422 on HTTP but silently degrade on websocket. Blank/whitespace values are normalized to absent.
2. **`TaskCreateRequest.timezone` is inert on the plain-agent branch.** `create_public_chat_task` is a dispatcher: workforce requests are delegated, everything else returns without starting a turn, and the frontend then calls `sendMessage` (which #1712 covers). All three handlers share one request model, so the field is accepted there but unused. Adding a separate model to hide it would cost more than it buys.
3. **`create_pending_workforce_run` does not take the parameter.** It creates a committed PENDING record for a caller-owned dispatch phase and starts no turn, so a timezone there would be dead weight.
4. **`patch_schedule_bg` in `tests/web/api/conftest.py` now returns the captured kwargs.** Its docstring records that three suites previously kept drifting private copies, so the shared helper was extended rather than shadowed. Existing callers ignore the return value and are unaffected.
5. **`resolveReportedTimezone` is now exported** from `use-websocket.ts` so the opening-message path reuses the exact resolution order (`data-timezone` → browser → omitted) instead of growing a second implementation that would drift.

## Explicit non-goals

- This PR does not add an account-level timezone setting.
- This PR does not wire scheduled triggers, which store a zone with different semantics ("the zone the schedule is expressed in").
- This PR does not address #1676.
- Adjacent pre-existing issues should be tracked as follow-ups rather than implemented here.

## Blocking criteria

Block only on a regression, on the stated invariant still being breakable, or on materially worsening a pre-existing problem.

## Tests

9 new tests across four suites; two existing frontend contract assertions updated. Own increment is 88% tests.

Mutation-verified, all six caught:

| Mutation | Result |
| --- | --- |
| `workforce_runs` stops passing `context` | 3 failed |
| widget + share handlers stop passing `timezone` | 2 failed |
| SDK stops passing `context` | 1 failed |
| blank timezone no longer normalized to absent | 1 failed |
| frontend stops putting the zone on `taskPayload` | 2 failed |
| frontend sends an empty zone instead of omitting | 1 failed |

Coverage: widget opening turn · share opening turn · SDK create · service-level forwarding · blank-zone normalization · frontend opening body carries the zone while still not calling `sendMessage` · frontend omits the key when the browser resolves none.

221 tests pass across the four touched web suites; 341 across `src/hooks` and `src/components/widget`. `ruff`, `mypy`, `tsc`, and `eslint` clean on the touched files.

The two frontend contract assertions now pin `Intl` to `Australia/Melbourne` in `beforeEach` rather than asserting the host zone — the same tautology trap @gemini-code-assist caught in #1712.

### CODE REVIEW

File: src/xagent/web/services/workforce_runs.py [hunk 1]
```diff
@@ -753,6 +757,7 @@ async def create_workforce_run_by_id(
     source: str | None = None,
     idempotency_key: str | None = None,
     extra_agent_config: dict[str, Any] | None = None,
+    timezone: str | None = None,
```
Comment (rogercloud) [inline src/xagent/web/services/workforce_runs.py:761]: **Major — Blocking: yes.** This adds the service parameter, but three supported workforce create-first callers still cannot populate it: authenticated `POST /api/workforces/{id}/runs`, `/api/workforces/preview/runs`, and `POST /v1/workforces/{id}/runs`. Their first turn is claimed and scheduled inside HTTP, so no later timezone-bearing WebSocket frame can repair a UTC relative-date answer (and resending would execute it twice). Please add an optional timezone to each request/payload contract, pass it through each handler and the preview service, and add seam coverage for value/blank/absent; keep fail-soft behavior and do not resend over WebSocket.

File: frontend/public/widget.js [hunk 2]
```diff
@@ -209,6 +209,16 @@
   var iconColor = scriptTag.getAttribute('data-icon-color') || '#fff';
   var panelBgColor = scriptTag.getAttribute('data-panel-bg-color') || '#fff';
 
+  // Read by use-websocket.ts off the iframe URL, where it outranks the browser
+  // zone. Optional: absent means the iframe URL is unchanged.
+  var embedTimezone = (scriptTag.getAttribute('data-timezone') || '').trim();
```
Comment (rogercloud) [inline frontend/public/widget.js:214]: **Minor — Blocking: no.** `data-timezone` now works, but the generated snippet and both agent/workforce widget dialogs provide no customer-facing way to discover this manual override. An embedder whose facility zone differs from the admin/browser therefore falls back to the browser zone unless it already knows to edit the tag. Please add a short UI/integration note with `data-timezone="Australia/Perth"`, the declared → browser/`Intl` → UTC precedence, blank omission, and invalid-IANA UTC fallback; keep the generated snippet free of an admin timezone.

**Review by rogercloud — CHANGES_REQUESTED** [review]
## Summary

This PR carries an optional caller timezone into turns that begin during task creation and into normal chat payloads: browser/`data-timezone` resolution feeds widget opening requests, public/share and `/v1/chat/tasks` handlers pass it to the existing scheduler context, and logical WebSocket retries retain a stable attempt context. It preserves omission/fail-soft UTC behavior and intentionally avoids resending workforce opening messages over WebSocket. Three supported workforce create-first paths still do not carry the field, and the new public widget override has no customer-facing discovery path. Blocking: yes — recommended event: REQUEST_CHANGES

## Update since the last review

Five commits are present on this head: `e877bc44` (browser timezone on chat frames), `d12c397b` (deterministic timezone assertions), `4d385b0d` (logical retry binding), `e8593400` (`data-timezone` widget override), and `3998ffe8` (task-creation forwarding). Together they extend timezone transport to the covered widget/share and `/v1/chat/tasks` openers while preserving fail-soft omission and no-resend behavior. No substantive prior #1726 findings existed; the only conversation item was Gemini's quota warning (`5420899915`), which was ignored.

## Approach verdict

**Acceptable with reservations.** Reusing the existing `context` → execution metadata → renderer pipeline, while binding the value to the logical message ID for retries, is the smallest coherent design: the new request fields are additive, old callers remain compatible, and no new persistence or transport mechanism is needed. The reservation is an incomplete enumeration of the create-first workforce boundaries: the service parameter was added without migrating three live workforce contracts. This is design-level **D1**, and its concrete impact and remediation are reported once below as **W1**, not as a second finding.

### D1 — Incomplete workforce create-first coverage (design-level)

The transport approach is sound, but the API boundary is not complete as delivered. Saved-workforce Web runs, unsaved previews, and the workforce SDK each start their first turn inside the HTTP request; leaving those contracts without a timezone means the central service parameter cannot fulfill the PR's stated transport invariant on those surfaces.

## Prior-findings checklist

- [x] **No substantive prior #1726 roots.** The complete paginated review-body and inline exports are `[[]]`; the conversation export contains only Gemini Code Assist quota warning comment `5420899915`, with no substantive finding, reply chain, waiver, or tracking reference.
- [x] **Quota comment ignored.** Comment `5420899915` is operational quota status, not review feedback, and is not counted as a root.
- [x] **Linked context is not #1726 history.** Findings and replies from #1712 and #1713 were read to establish scope only; they are not prior findings for this PR.

## Findings by severity

### Major

#### W1 — Workforce create-first entry points still lose timezone

**Anchor:** `src/xagent/web/services/workforce_runs.py:760`  
**Severity:** major  
**Blocking:** yes  
**Source:** [new]

**Trigger:** A user starts the first message through any of these supported HTTP-first workforce paths: the authenticated Web saved-workforce run (`POST /api/workforces/{id}/runs`), an unsaved workforce builder preview (`POST /api/workforces/preview/runs`), or the workforce SDK (`POST /v1/workforces/{id}/runs`), during a local/UTC date-boundary window (for example, UTC+10 between local midnight and 10:00) with a relative-date question such as “tomorrow.”

**Impact:** These requests claim and schedule the opening turn inside HTTP. Because none carries a timezone, that first answer still uses UTC (or the renderer's UTC fallback), so it can select the wrong calendar date. A later WebSocket turn cannot repair an already returned answer, and resending the opener over WebSocket would execute it twice.

**Evidence:** The PR adds `timezone` at `workforce_runs.py:760` and schedules `context={"timezone": request.timezone}` at `:733`, but the authenticated Web `WorkforceRunRequest` and handler (`src/xagent/web/api/workforces.py:125-131,1407-1423`) do not define or pass it. The preview request/handler and service likewise have no field or context (`src/xagent/web/api/workforces.py:133-141,644-666`; `src/xagent/web/services/workforce_runs.py:828-885`). The SDK request model and adapter omit it (`src/xagent/web/schemas/v1.py:468-498`; `src/xagent/web/api/v1/workforces.py:132-145`), and the frontend payloads/openers are unchanged (`frontend/src/types/workforce.ts:196-223`, `frontend/src/app/workforces/[id]/run/page-client.tsx:645-681`, `frontend/src/components/workforce/workforce-builder.tsx:543-574`). The full diff confirms the service additions are present while these callers are outside the changed surface. The target base already contains the paired timezone-aware renderer from #1709; because this is a stacked branch, the review worktree HEAD itself still shows the pre-#1709 UTC implementation at `src/xagent/core/agent/context/execution.py:444-450` and `src/xagent/core/agent/pattern/react/react.py:840-867`. In the integrated target, missing metadata still selects that renderer's UTC fallback.

**Specific fix:** Add the same optional top-level `timezone` field to `WorkforceRunRequest`, `WorkforcePreviewRunRequest`, and `CreateWorkforceRunRequest` (not hidden inside the SDK `MessageBody`), plus `WorkforceRunPayload` and `WorkforcePreviewRunPayload`. Resolve it with the existing `resolveReportedTimezone()` in both authenticated Web openers and the builder, pass it through each handler, extend `create_preview_workforce_run` and its normalization to schedule the same context, and forward it from the SDK adapter. Add discriminating seam/API tests for each real entry point (value reaches scheduler; absent/blank produces `None`), preserve fail-soft invalid-IANA behavior, and do not resend the opener over WebSocket.

### Minor

#### W2 — Public `data-timezone` override is not discoverable

**Anchor:** `frontend/public/widget.js:214`  
**Severity:** minor  
**Blocking:** no  
**Source:** [new]

**Trigger:** An embedder needs a facility/business timezone that differs from the administrator's or end user's browser timezone, but uses the product-generated widget snippet without already knowing the new manual attribute.

**Impact:** The implementation works, but the customer has no visible path to discover `data-timezone`; the embed therefore silently uses the browser zone unless the customer reads source or PR material. In remote-staff/fleet deployments this can leave relative-date answers based on the wrong machine zone, but the omission is a usability/documentation gap rather than a blocker because the manual override and browser fallback are functional.

**Evidence:** `frontend/public/widget.js:214-220` trims and URL-encodes the optional value for both iframe URL shapes, and `resolveReportedTimezone()` gives it precedence over `Intl`; invalid values can still fail soft to UTC. However, `frontend/src/lib/agent-widget-config.ts:9-18` continues to generate only the existing attributes, while the agent and workforce copy-snippet surfaces merely say to copy the script (`frontend/src/components/build/deploy-agent-dialog.tsx:647-649`, `frontend/src/components/workforce/workforce-widget-dialog.tsx:304-306`). The locale strings and repository docs contain no manual override guidance. This is a consequence of the new public API in this PR; the related #1713 scope discussion is context, not prior #1726 history.

**Specific fix:** Add a short customer-facing hint beside both generated snippets and/or a stable widget integration guide. Show an optional `data-timezone="Australia/Perth"` IANA example, document declared → browser/`Intl` → UTC precedence, state that blank/absent omits the override and invalid IANA values fall back to UTC, and keep the generated snippet free of an administrator's machine timezone.

## Review notes

No local tests, builds, linters, or formatters were run for this review. The Simplification Lens was unavailable because of `usage_limit_reached`; no simplification claims are included and no Simplification section is emitted.

## Blocking status & recommended decision

Blocking: yes

Blocking issues:
- **W1** — `src/xagent/web/services/workforce_runs.py:760` — **major** — authenticated Web, preview, and workforce SDK opening turns can still run without timezone and answer relative dates against UTC. **[new]**

Non-blocking findings:
- **W2** — `frontend/public/widget.js:214` — **minor**, **Blocking: no** — the manual embed timezone override is functional but undiscoverable. **[new]**

Blocking: yes — recommended event: REQUEST_CHANGES

**Comment by bluefish-08** [comment]
W1 (major) fixed in the latest push; W2 (docs) is in progress.

**W1 — workforce create-first paths lose timezone.** Confirmed all three start their opening turn inside the HTTP request with no timezone. Fixed by carrying an optional `timezone` through each:

- `WorkforceRunRequest`, `WorkforcePreviewRunRequest`, and the SDK `CreateWorkforceRunRequest` each gain a top-level optional `timezone` (not hidden in `MessageBody`).
- The authenticated Web handler and the SDK adapter pass it to `create_workforce_run` / `create_workforce_run_by_id` (which already accepted the param from the paired PR); `create_preview_workforce_run` gained the param and schedules `context={"timezone": ...}` the same way.
- Frontend: `runWorkforce` and `runWorkforcePreview` now attach `resolveReportedTimezone()` centrally (an explicit payload value wins), so both the saved-run opener and the builder preview opener carry it without per-caller changes.
- Fail-soft preserved: blank normalizes to absent, unresolvable IANA degrades to UTC at the render site, no opener is resent over the websocket.

Contract tests for all three entry points (value reaches the scheduler; absent/blank → `None`), plus the frontend helper injection and explicit-override cases. Mutation-verified: dropping the forward on each path fails its test (Web 2, SDK 1, preview 1; frontend helper 2).

**W2 — `data-timezone` discoverability.** Same finding as #1713 Group C; being addressed on the #1713 branch with a localized hint beside both copy-snippet surfaces, keeping the generated snippet free of a default.

This PR is stacked on #1713, which is stacked on #1712; both have been rebased onto their updated parents.

File: src/xagent/web/services/workforce_runs.py (hunk 1 above)
Comment (bluefish-08) [reply src/xagent/web/services/workforce_runs.py:761]: Fixed in the latest push. All three create-first paths (authenticated Web run, builder preview, workforce SDK) now carry an optional top-level timezone into the scheduler context; frontend openers attach it via resolveReportedTimezone. Contract tests per path, mutation-verified.

File: frontend/public/widget.js (hunk 2 above)
Comment (bluefish-08) [reply frontend/public/widget.js:214]: Addressed on the #1713 branch (same finding as its Group C): a localized hint now sits beside both copy-snippet surfaces (agent deploy dialog and workforce widget dialog) with the IANA example and the data-timezone → browser → UTC precedence. buildWidgetSnippet still emits no timezone default. Keeping it in #1713 since that PR owns the data-timezone attribute.

File: src/xagent/web/api/v1/tasks.py [hunk 3]
```diff
@@ -633,6 +633,7 @@ async def _prepare_and_schedule() -> tuple[_PreparedCreateTaskStart, TurnStarted
                 actor_user_id=actor_user_id,
                 payload=prepared.payload,
                 claimed=prepared.claimed_turn,
+                context=({"timezone": request.timezone} if request.timezone else None),
```
Comment (rogercloud) [inline src/xagent/web/api/v1/tasks.py]: **Minor, Blocking: no.** This builds `context=({"timezone": request.timezone} if request.timezone else None)` using raw truthiness with no `.strip()`. The workforce path (`workforce_runs.py:196-200`, `_normalize_workforce_run_request`) normalizes with `(timezone or "").strip() or None`, so a whitespace-only string like `"   "` is dropped there but passed through verbatim here. The downstream `clock_zone()` consumer already strips and degrades to UTC on empty input, so this is a data-hygiene inconsistency rather than a crash risk - but it makes the PR's stated "blank values normalized to absent" invariant inaccurate for this entry point. Suggest `(request.timezone or "").strip() or None` here, ideally factored into a small shared helper used by all three `context={"timezone": ...}` call sites (this file, and `workforce_runs.py:733,887`).

File: frontend/src/lib/workforces-api.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ import {
   parseApiResponse,
 } from "@/lib/api-wrapper"
 import { getApiUrl } from "@/lib/utils"
+import { resolveReportedTimezone } from "@/hooks/use-websocket"
```
Comment (rogercloud) [inline frontend/src/lib/workforces-api.ts:10]: **Minor, Blocking: no.** This module now imports `resolveReportedTimezone` from `hooks/use-websocket.ts`, a large (~1500-line) hooks module that also pulls in `react` and `@/contexts/auth-context`, even though `resolveReportedTimezone` itself is a pure function (only reads `window.location.search` and `Intl`). All current importers of this file are already client components, so there's no SSR-boundary issue, but the coupling is unnecessary and forces the unit test to mock the entire hooks module. Suggest extracting `resolveReportedTimezone` into a standalone `lib/timezone.ts` with no React/auth dependencies, imported by both `use-websocket.ts` and this file.

File: src/xagent/web/schemas/chat.py [hunk 5]
```diff
@@ -85,6 +85,10 @@ class TaskCreateRequest(BaseModel):
     )
     is_preview: bool = False  # Backward-compatible alias for is_visible=False.
     is_visible: bool = True
+    # Only consumed by entry points that start the first turn inside task
+    # creation. Unresolvable names degrade to UTC at the render site rather
+    # than 422 here, matching how the websocket path treats the same field.
+    timezone: Optional[str] = None  # IANA name for the caller's local clock
```
Comment (rogercloud) [inline src/xagent/web/schemas/chat.py]: **Low, Blocking: no.** The new `timezone` field has no `max_length`, unlike `idempotency_key` in the same schemas (`max_length=128`). These fields are reachable by unauthenticated guests via widget/share-link create endpoints. Low risk since the same requests already carry other unbounded free-text fields (`message`, `description`, etc.), but suggest adding `max_length=64` (consistent with actual IANA zone-name lengths) as minor hardening, and applying the same to the sibling `timezone` fields in `schemas/v1.py` and `api/workforces.py:134,145`.

File: src/xagent/web/services/workforce_runs.py [hunk 6]
```diff
@@ -195,6 +197,7 @@ def _normalize_workforce_run_request(
         extra_agent_config=(
             dict(extra_agent_config) if extra_agent_config is not None else None
         ),
+        timezone=(timezone or "").strip() or None,
```
Comment (rogercloud) [inline src/xagent/web/services/workforce_runs.py]: **Low, Blocking: no.** `_normalize_workforce_run_request` does `(timezone or "").strip()` unconditionally - a non-string value would raise `AttributeError` instead of a typed error, inconsistent with sibling normalizers in this file (`_normalize_selected_file_ids` skips non-str elements silently; `_normalize_idempotency_key` raises a typed `WorkforceRunError`). All current callers pass Pydantic-validated `Optional[str]`, so this isn't reachable today, but suggest adding an `isinstance` guard for consistency with the established pattern.

File: src/xagent/web/api/v1/workforces.py [hunk 7]
```diff
@@ -142,6 +142,7 @@ async def create_workforce_run_endpoint(
             is_visible=False,
             source="sdk",
             idempotency_key=request.idempotency_key,
+            timezone=request.timezone,
```
Comment (rogercloud) [inline src/xagent/web/api/v1/workforces.py:145]: **Low, Blocking: no.** `timezone` is passed alongside `idempotency_key`, but the idempotency lookup (`workforce_runs.py:200-234`) is keyed only on `workforce_id`+`idempotency_key`. On a replay with the same key but a different `timezone`, the original run's timezone is silently kept. This is a pre-existing property of the shared idempotency-replay mechanism (every other field is equally ignored on replay, and `created: bool` in the response lets a caller detect a replay), not something this PR introduced - flagging as a separate, non-blocking improvement to the shared mechanism rather than a defect here.

File: frontend/src/hooks/use-websocket.ts [hunk 8]
```diff
@@ -67,6 +67,22 @@ export class MessageDeliveryError extends Error {
   }
 }
 
+export const resolveReportedTimezone = (): string | undefined => {
```
Comment (rogercloud) [inline frontend/src/hooks/use-websocket.ts:80]: **Low, Blocking: no.** `resolveReportedTimezone` resolves URL `?timezone=` before falling back to the browser's `Intl` zone. It was originally exercised mainly through the embeddable widget iframe (host-controlled URL), but is now also used unconditionally from the main authenticated app's `runWorkforce`/`runWorkforcePreview` calls. A stray or shared `?timezone=` param on a bookmarked/campaign URL in the main app would silently override the real browser timezone for that run. Impact is limited to the "current time" prompt hint, with no auth/injection concern, but there's no test covering this URL-param branch through the new call paths. Suggest namespacing the query param and adding a regression test.

**Review by rogercloud — APPROVED** [review]
## PR Summary

PR #1726 is the final increment of a 3-PR stack (#1712 -> #1713 -> #1726) fixing issue #1675, where the agent runtime always injected "current time" into prompts as UTC because no timezone was available anywhere in the system to inject instead. #1712/#1713 (separate, still-open PRs) add browser-timezone reporting over the websocket `sendMessage` path and a widget `data-timezone` embed attribute. This PR's own increment (commits be265bb8 and 6fa91716) wires that same timezone value into the two entry points that never go through the websocket path: workforce "opener" turns that start synchronously inside the HTTP task-creation request (widget create, share-link create, internal workforce-run-creation service), and the SDK/REST `POST /v1/chat/tasks` path, which has no websocket at all. The mechanism is pure wiring - feeding the pre-existing `context` parameter on `schedule_claimed_create_turn` with `{"timezone": <value>}`, flowing unchanged into `ExecutionContext.metadata["timezone"]`, consumed by the already-merged `clock_zone()` renderer (#1709). No new abstractions, no schema/migration/DB changes, no changes to `TaskTurnPayload`/orchestrator internals/runner.

Note: as GitHub renders this diff against `main`, it includes ~7 commits inherited from the still-open, separately-reviewed stacked PRs #1712/#1713. This review scopes its findings to PR #1726's own increment (commits be265bb8, 6fa91716), consistent with how the author's own PR body frames its "own increment."

**Blocking: no - recommended event: APPROVE**

## Design Evaluation

The design is sound. The PR reuses an existing `context` passthrough mechanism and an already-merged consumer rather than introducing new abstractions, and correctly scopes itself to the two entry points that bypass the websocket path. Two non-blocking design-level observations (detailed as Findings 1 and 5 below, not double-counted here):

- **Duplicated context-construction with inconsistent blank-normalization** across the three call sites that build `context={"timezone": ...}` - only the workforce path normalizes whitespace-only values; this PR's new SDK/REST path does not. A shared helper (e.g., in `task_orchestrator.py`, reusing #1709's `CLOCK_TIMEZONE_METADATA_KEY` constant instead of the string literal `"timezone"` hardcoded three times) would remove both the duplication and the inconsistency.
- **Frontend dependency inversion**: a previously widget-scoped timezone-resolution function is now imported into a general-purpose API client module by reaching into a large hooks module, rather than being extracted into a standalone utility.

Also explicitly checked and found sound: no IANA validation at the schema layer is safe (the downstream `clock_zone()` consumer already catches zone-lookup errors and degrades to UTC); `TaskCreateRequest.timezone` being inert on the plain-agent dispatch branch is harmless (that branch never starts a turn synchronously); `create_workforce_run_record` correctly excludes the parameter since it starts no turn; all `schedule_claimed_create_turn` call sites were enumerated, and the one omission (`a2a.py:808`) is a non-issue since the A2A protocol carries no timezone-like field to plumb.

## Simplification Opportunities

Lean already.

## Blocking Status & Recommended Decision

**Blocking: no** - no confirmed finding meets the blocker criteria: all degrade gracefully (the downstream `clock_zone()` consumer strips and falls back to UTC on any malformed/empty input), all have safe fallbacks, and none produce an incorrect user-visible result or violate a data/state contract under any credible trigger.

**Recommended event: APPROVE**

**Comment by bluefish-08** [comment]
Thanks for the approve. Addressed the two non-blocking design observations (pushed 72a1ef1c; does not change the approved behavior):

**Finding 1 — duplicated context construction + inconsistent blank-normalization: fixed.** Added a shared `timezone_schedule_context(timezone)` helper in `task_orchestrator.py` that all three create-first sites now call (workforce widget/share, workforce SDK, and `/v1/chat/tasks`). It uses #1709's `CLOCK_TIMEZONE_METADATA_KEY` constant instead of the hardcoded `"timezone"` literal, and normalizes whitespace-only to `None` — which closes the real inconsistency you flagged: the SDK/REST path previously did not strip, so `"   "` would have reached the renderer as a blank zone. Added `test_create_task_treats_a_blank_timezone_as_absent` for that path; mutation-verified the strip. Existing per-path contract tests still pass.

**Finding 2 — frontend dependency inversion (`resolveReportedTimezone` imported from the hooks module into the API client): deferring as a follow-up.** Extracting it into a standalone utility is a pure-structure move that touches import wiring across the widget/hook/api-client boundary for no behavior change; I'd rather not reshuffle that on an already-approved PR. Tracking it as a follow-up. Happy to pull it into this PR instead if you'd prefer it not be deferred.

File: src/xagent/web/api/v1/tasks.py (hunk 3 above)
Comment (bluefish-08) [reply src/xagent/web/api/v1/tasks.py]: Fixed. All three create-first sites now call a shared timezone_schedule_context(timezone) helper in task_orchestrator.py that strips and normalizes whitespace-only to None (using #1709's CLOCK_TIMEZONE_METADATA_KEY constant, not the string literal). So the SDK/REST path now matches the workforce path, and the 'blank normalized to absent' invariant holds for every entry point. Added test_create_task_treats_a_blank_timezone_as_absent; mutation-verified the strip.

File: src/xagent/web/schemas/chat.py (hunk 5 above)
Comment (bluefish-08) [reply src/xagent/web/schemas/chat.py]: Fixed. Added max_length=64 to the timezone field on all four schemas (chat.py, v1.py x2, api/workforces.py x2), matching IANA zone-name lengths and consistent with idempotency_key's bound. Added test_create_task_rejects_an_over_long_timezone (422 on 65 chars); mutation-verified.

File: src/xagent/web/services/workforce_runs.py (hunk 6 above)
Comment (bluefish-08) [reply src/xagent/web/services/workforce_runs.py]: Fixed. _normalize_workforce_run_request now guards with isinstance: (timezone.strip() or None) if isinstance(timezone, str) else None, so a non-string degrades to None instead of raising AttributeError — consistent with the sibling normalizers. Not reachable today (Pydantic gives Optional[str]) but matches the established pattern.

File: frontend/src/lib/workforces-api.ts (hunk 4 above)
Comment (bluefish-08) [reply frontend/src/lib/workforces-api.ts:10]: Deferring as a follow-up (same as the design-observation Finding 2 in your approve). Extracting resolveReportedTimezone into a standalone lib/timezone.ts is a pure-structure move touching import wiring across the widget/hook/api-client boundary for no behavior change; I'd rather not reshuffle that on an already-approved PR. Tracked as a follow-up.

File: src/xagent/web/api/v1/workforces.py (hunk 7 above)
Comment (bluefish-08) [reply src/xagent/web/api/v1/workforces.py:145]: Acknowledged — as you note, this is a pre-existing property of the shared idempotency-replay mechanism (every field is ignored on replay; created:false signals a replay), not introduced here. Leaving the shared mechanism as-is; not changing it in this PR.

File: frontend/src/hooks/use-websocket.ts (hunk 8 above)
Comment (bluefish-08) [reply frontend/src/hooks/use-websocket.ts:80]: Acknowledged, deferring. The ?timezone= URL-param precedence lives in resolveReportedTimezone (merged with #1713); namespacing it would touch both the widget.js writer and the reader across already-merged code, and the impact is limited to the current-time prompt hint with no auth/injection concern. Tracking as a follow-up together with YO4H (both are about that same shared function).
